nixos/modules/services/desktops/pipewire.nix

```diff
@@ -25,17 +27,74 @@ in {
           Automatically run pipewire when connections are made to the pipewire socket.
         '';
       };
+
+      alsa = {
+        enable = mkOption {
+          default = false;
+          type = types.bool;
+          description = ''
+            Route audio to/from generic ALSA-using applications via the ALSA PIPEWIRE PCM plugin.
+          '';
+        };
+
+        support32Bit = mkOption {
+          default = false;
+          type = types.bool;
+          description = ''
+            Whether to support sound for 32-bit ALSA applications on a 64-bit system.
+          '';
+        };
+      };
+
+      jack = {
+        enable = mkOption {
+          default = false;
+          type = types.bool;
+          description = ''
+            Enable transparent JACK audio emulation using LD_LIBRARY_PATH.
+          '';
+        };
+      };
+
+      pulse = {
+        enable = mkOption {
+          default = false;
+          type = types.bool;
+          description = ''
+            Enable transparent PulseAudio emulation using LD_LIBRARY_PATH.
+          '';
+        };
+      };
     };
   };
 
 
   ###### implementation
   config = mkIf cfg.enable {
-    environment.systemPackages = packages;
+    environment.systemPackages = [ pkgs.pipewire ];
 
-    systemd.packages = packages;
+    systemd.packages = [ pkgs.pipewire ];
 
     systemd.user.sockets.pipewire.wantedBy = lib.mkIf cfg.socketActivation [ "sockets.target" ];
-  };
 
+    sound.extraConfig = mkIf cfg.alsa.enable ''
+      pcm_type.pipewire {
+        libs.native = ${pkgs.pipewire.lib}/lib/alsa-lib/libasound_module_pcm_pipewire.so ;
+        ${optionalString enable32BitAlsaPlugins
+          "libs.32Bit = ${pkgs.pkgsi686Linux.pipewire.lib}/lib/alsa-lib/libasound_module_pcm_pipewire.so ;"}
+      }
+      pcm.!default {
+        @func getenv
+        vars [ PCM ]
+        default "plug:pipewire"
+        playback_mode "-1"
+        capture_mode "-1"
+      }
+    '';
+    environment.etc."alsa/conf.d/50-pipewire.conf".source = mkIf cfg.alsa.enable
+      "${pkgs.pipewire}/share/alsa/alsa.conf.d/50-pipewire.conf";
+    environment.sessionVariables.LD_LIBRARY_PATH =
+      (optional cfg.jack.enable "${pkgs.pipewire.lib}/lib/pipewire-0.3/jack") ++
+      (optional cfg.pulse.enable "${pkgs.pipewire.lib}/lib/pipewire-0.3/pulse");
```

Comment:
Another way to do this is to link those into the systemPackages in the lib/ folder as glibc will automatically look for libraries there. This avoids the issue that users have to log out and back in for these changes to take effect.

Reply:
I tried to do this earlier, however it didn't seem to work. I believe that this is because programs manually specify dependencies on PulseAudio and JACK. Because of this they refer to the libraries by absolute path where they are successfully found before checking the default path. This means that programs end up using the original libraries instead of the PipeWire wrappers.

In other words:
`LD_LIBRARY_PATH` > hard-coded paths > /run/current-system/sw/lib